larpandoracontent/LArReclustering/ThreeDReclusteringAlgorithm.cc

```diff
@@ -0,0 +1,483 @@
+            if (!parameters.m_caloHitList.empty())
+            {
+                PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::Cluster::Create(*this, parameters, pNewTwoDCluster));
+            }
+            if(pNewTwoDCluster!=nullptr && !parameters.m_caloHitList.empty() && hitType==TPC_VIEW_U) {m_newClustersUMap.insert(std::make_pair(iCluster,pNewTwoDCluster));}
+            else if(pNewTwoDCluster!=nullptr && !parameters.m_caloHitList.empty() && hitType==TPC_VIEW_V) {m_newClustersVMap.insert(std::make_pair(iCluster,pNewTwoDCluster));}
+            else if(pNewTwoDCluster!=nullptr && !parameters.m_caloHitList.empty() && hitType==TPC_VIEW_W) {m_newClustersWMap.insert(std::make_pair(iCluster,pNewTwoDCluster));}
+            
+            iCluster++;
+        }
+
+        //Check the leftover caloHits. Attach to the nearest cluster in the new cluster list.
+        std::map<int,const Cluster*> clustersForLeftoverHitsMap;
+        if(hitType==TPC_VIEW_U) clustersForLeftoverHitsMap = m_newClustersUMap;
+        else if(hitType==TPC_VIEW_V) clustersForLeftoverHitsMap = m_newClustersVMap;
+        else if(hitType==TPC_VIEW_W) clustersForLeftoverHitsMap = m_newClustersWMap;
+        if(clustersForLeftoverHitsMap.size())
+        {
+        for(const OrderedCaloHitList::value_type &mapEntry : leftoverCaloHitList)
+            {
+                for (const CaloHit *const pCaloHit : *mapEntry.second)
+                {
+                    const Cluster* pNearestCluster(nullptr);
+                    double minimumDistance(std::numeric_limits<float>::max());
+                    for(const auto & [clusterIndex, pNewTwoDCluster] : clustersForLeftoverHitsMap)
+                    {
+                        double dist = LArClusterHelper::GetClosestDistance(pCaloHit->GetPositionVector(), pNewTwoDCluster);  
+                        if (dist<minimumDistance)
+                        {
+                            minimumDistance=dist;
+                            pNearestCluster=pNewTwoDCluster;
+                        }
+                    }
+                    PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::AddToCluster(*this,pNearestCluster,pCaloHit));
+                }
+            }
+        }
+
+        PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::EndFragmentation(*this, fragmentListName, originalListName));
+        PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::ReplaceCurrentList<Cluster>(*this, initialListName));
+    }
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode ThreeDReclusteringAlgorithm::BuildNewPfos(const Pfo *pPfoToRebuild, ClusterList &newClustersList)
+{
+    const PfoList *pNewPfoList(nullptr);
+    std::string newPfoListName = "changedShowerParticles3D";
```

Comment:
Configurable?

Reply:
This is also just an internal list, so not worth it. There may be a way to make it so that all these internal list names don't have to be defined, I suspect I was running into some limitations of the API functions that required these.